diff --git a/plugins/format/address.inc b/plugins/format/address.inc
index d5426ad..8f7caa4 100644
--- a/plugins/format/address.inc
+++ b/plugins/format/address.inc
@@ -122,73 +122,12 @@ function addressfield_format_address_generate(&$format, $address, $context = arr
   }
 
   // A few countries have a well-known list of administrative divisions.
+  $countries_with_administrative_divisions = array('US', 'IT', 'BR', 'CA', 'AU', 'NZ');
+  if (in_array($address['country'], $countries_with_administrative_divisions)) {
+    $format['locality_block']['administrative_area']['#options'] = country_administrative_areas($address['country']);
+  }
+
   if ($address['country'] == 'US') {
-    $format['locality_block']['administrative_area']['#options'] = array(
-      ''   => t('--'),
-      'AL' => t('Alabama'),
-      'AK' => t('Alaska'),
-      'AZ' => t('Arizona'),
-      'AR' => t('Arkansas'),
-      'CA' => t('California'),
-      'CO' => t('Colorado'),
-      'CT' => t('Connecticut'),
-      'DE' => t('Delaware'),
-      'DC' => t('District Of Columbia'),
-      'FL' => t('Florida'),
-      'GA' => t('Georgia'),
-      'HI' => t('Hawaii'),
-      'ID' => t('Idaho'),
-      'IL' => t('Illinois'),
-      'IN' => t('Indiana'),
-      'IA' => t('Iowa'),
-      'KS' => t('Kansas'),
-      'KY' => t('Kentucky'),
-      'LA' => t('Louisiana'),
-      'ME' => t('Maine'),
-      'MD' => t('Maryland'),
-      'MA' => t('Massachusetts'),
-      'MI' => t('Michigan'),
-      'MN' => t('Minnesota'),
-      'MS' => t('Mississippi'),
-      'MO' => t('Missouri'),
-      'MT' => t('Montana'),
-      'NE' => t('Nebraska'),
-      'NV' => t('Nevada'),
-      'NH' => t('New Hampshire'),
-      'NJ' => t('New Jersey'),
-      'NM' => t('New Mexico'),
-      'NY' => t('New York'),
-      'NC' => t('North Carolina'),
-      'ND' => t('North Dakota'),
-      'OH' => t('Ohio'),
-      'OK' => t('Oklahoma'),
-      'OR' => t('Oregon'),
-      'PA' => t('Pennsylvania'),
-      'RI' => t('Rhode Island'),
-      'SC' => t('South Carolina'),
-      'SD' => t('South Dakota'),
-      'TN' => t('Tennessee'),
-      'TX' => t('Texas'),
-      'UT' => t('Utah'),
-      'VT' => t('Vermont'),
-      'VA' => t('Virginia'),
-      'WA' => t('Washington'),
-      'WV' => t('West Virginia'),
-      'WI' => t('Wisconsin'),
-      'WY' => t('Wyoming'),
-      ' ' => t('--'),
-      'AA' => t('Armed Forces (Americas)'),
-      'AE' => t('Armed Forces (Europe, Canada, Middle East, Africa)'),
-      'AP' => t('Armed Forces (Pacific)'),
-      'AS' => t('American Samoa'),
-      'FM' => t('Federated States of Micronesia'),
-      'GU' => t('Guam'),
-      'MH' => t('Marshall Islands'),
-      'MP' => t('Northern Mariana Islands'),
-      'PW' => t('Palau'),
-      'PR' => t('Puerto Rico'),
-      'VI' => t('Virgin Islands'),
-    );
     $format['locality_block']['postal_code']['#title'] = t('ZIP Code');
 
     if ($context['mode'] == 'render') {
@@ -196,119 +135,6 @@ function addressfield_format_address_generate(&$format, $address, $context = arr
     }
   }
   else if ($address['country'] == 'IT') {
-    $format['locality_block']['administrative_area']['#options'] = array(
-      '' => t('--'),
-      'AG' => t('Agrigento'),
-      'AL' => t('Alessandria'),
-      'AN' => t('Ancona'),
-      'AO' => t("Valle d'Aosta/Vallée d'Aoste"),
-      'AP' => t('Ascoli Piceno'),
-      'AQ' => t("L'Aquila"),
-      'AR' => t('Arezzo'),
-      'AT' => t('Asti'),
-      'AV' => t('Avellino'),
-      'BA' => t('Bari'),
-      'BG' => t('Bergamo'),
-      'BI' => t('Biella'),
-      'BL' => t('Belluno'),
-      'BN' => t('Benevento'),
-      'BO' => t('Bologna'),
-      'BR' => t('Brindisi'),
-      'BS' => t('Brescia'),
-      'BT' => t('Barletta-Andria-Trani'),
-      'BZ' => t('Bolzano/Bozen'),
-      'CA' => t('Cagliari'),
-      'CB' => t('Campobasso'),
-      'CE' => t('Caserta'),
-      'CH' => t('Chieti'),
-      'CI' => t('Carbonia-Iglesias'),
-      'CL' => t('Caltanissetta'),
-      'CN' => t('Cuneo'),
-      'CO' => t('Como'),
-      'CR' => t('Cremona'),
-      'CS' => t('Cosenza'),
-      'CT' => t('Catania'),
-      'CZ' => t('Catanzaro'),
-      'EN' => t('Enna'),
-      'FC' => t('Forlì-Cesena'),
-      'FE' => t('Ferrara'),
-      'FG' => t('Foggia'),
-      'FI' => t('Firenze'),
-      'FM' => t('Fermo'),
-      'FR' => t('Frosinone'),
-      'GE' => t('Genova'),
-      'GO' => t('Gorizia'),
-      'GR' => t('Grosseto'),
-      'IM' => t('Imperia'),
-      'IS' => t('Isernia'),
-      'KR' => t('Crotone'),
-      'LC' => t('Lecco'),
-      'LE' => t('Lecce'),
-      'LI' => t('Livorno'),
-      'LO' => t('Lodi'),
-      'LT' => t('Latina'),
-      'LU' => t('Lucca'),
-      'MB' => t('Monza e Brianza'),
-      'MC' => t('Macerata'),
-      'ME' => t('Messina'),
-      'MI' => t('Milano'),
-      'MN' => t('Mantova'),
-      'MO' => t('Modena'),
-      'MS' => t('Massa-Carrara'),
-      'MT' => t('Matera'),
-      'NA' => t('Napoli'),
-      'NO' => t('Novara'),
-      'NU' => t('Nuoro'),
-      'OG' => t('Ogliastra'),
-      'OR' => t('Oristano'),
-      'OT' => t('Olbia-Tempio'),
-      'PA' => t('Palermo'),
-      'PC' => t('Piacenza'),
-      'PD' => t('Padova'),
-      'PE' => t('Pescara'),
-      'PG' => t('Perugia'),
-      'PI' => t('Pisa'),
-      'PN' => t('Pordenone'),
-      'PO' => t('Prato'),
-      'PR' => t('Parma'),
-      'PT' => t('Pistoia'),
-      'PU' => t('Pesaro e Urbino'),
-      'PV' => t('Pavia'),
-      'PZ' => t('Potenza'),
-      'RA' => t('Ravenna'),
-      'RC' => t('Reggio di Calabria'),
-      'RE' => t("Reggio nell'Emilia"),
-      'RG' => t('Ragusa'),
-      'RI' => t('Rieti'),
-      'RM' => t('Roma'),
-      'RN' => t('Rimini'),
-      'RO' => t('Rovigo'),
-      'SA' => t('Salerno'),
-      'SI' => t('Siena'),
-      'SO' => t('Sondrio'),
-      'SP' => t('La Spezia'),
-      'SR' => t('Siracusa'),
-      'SS' => t('Sassari'),
-      'SV' => t('Savona'),
-      'TA' => t('Taranto'),
-      'TE' => t('Teramo'),
-      'TN' => t('Trento'),
-      'TO' => t('Torino'),
-      'TP' => t('Trapani'),
-      'TR' => t('Terni'),
-      'TS' => t('Trieste'),
-      'TV' => t('Treviso'),
-      'UD' => t('Udine'),
-      'VA' => t('Varese'),
-      'VB' => t('Verbano-Cusio-Ossola'),
-      'VC' => t('Vercelli'),
-      'VE' => t('Venezia'),
-      'VI' => t('Vicenza'),
-      'VR' => t('Verona'),
-      'VS' => t('Medio Campidano'),
-      'VT' => t('Viterbo'),
-      'VV' => t('Vibo Valentia'),
-    );
     $format['locality_block']['administrative_area']['#title'] = t('Province');
   }
   elseif ($address['country'] == 'BR') {
@@ -320,36 +146,6 @@ function addressfield_format_address_generate(&$format, $address, $context = arr
       '#render_option_value' => FALSE,
     );
     $format['locality_block']['administrative_area']['#render_option_value'] = TRUE;
-    $format['locality_block']['administrative_area']['#options'] = array(
-      '' => t('--'),
-      'AC' => t('Acre'),
-      'AL' => t('Alagoas'),
-      'AM' => t('Amazonas'),
-      'AP' => t('Amapá'),
-      'BA' => t('Bahia'),
-      'CE' => t('Ceará'),
-      'DF' => t('Distrito Federal'),
-      'ES' => t('Espírito Santo'),
-      'GO' => t('Goiás'),
-      'MA' => t('Maranhão'),
-      'MG' => t('Minas Gerais'),
-      'MS' => t('Mato Grosso do Sul'),
-      'MT' => t('Mato Grosso'),
-      'PA' => t('Pará'),
-      'PB' => t('Paraíba'),
-      'PE' => t('Pernambuco'),
-      'PI' => t('Piauí'),
-      'PR' => t('Paraná'),
-      'RJ' => t('Rio de Janeiro'),
-      'RN' => t('Rio Grande do Norte'),
-      'RO' => t('Rondônia'),
-      'RR' => t('Roraima'),
-      'RS' => t('Rio Grande do Sul'),
-      'SC' => t('Santa Catarina'),
-      'SE' => t('Sergipe'),
-      'SP' => t('São Paulo'),
-      'TO' => t('Tocantins'),
-    );
     // Change some titles to make translation easier.
     $format['street_block']['#attributes'] = array(
       'class' => array('addressfield-container-inline'),
@@ -390,22 +186,6 @@ function addressfield_format_address_generate(&$format, $address, $context = arr
     $format['locality_block']['postal_code']['#tag'] = 'div';
   }
   else if ($address['country'] == 'CA') {
-    $format['locality_block']['administrative_area']['#options'] = array(
-      '' => t('--'),
-      'AB' => t('Alberta'),
-      'BC' => t('British Columbia'),
-      'MB' => t('Manitoba'),
-      'NB' => t('New Brunswick'),
-      'NL' => t('Newfoundland and Labrador'),
-      'NT' => t('Northwest Territories'),
-      'NS' => t('Nova Scotia'),
-      'NU' => t('Nunavut'),
-      'ON' => t('Ontario'),
-      'PE' => t('Prince Edward Island'),
-      'QC' => t('Quebec'),
-      'SK' => t('Saskatchewan'),
-      'YT' => t('Yukon Territory'),
-    );
     $format['locality_block']['administrative_area']['#title'] = t('Province');
 
     if ($context['mode'] == 'render') {
@@ -413,17 +193,6 @@ function addressfield_format_address_generate(&$format, $address, $context = arr
     }
   }
   else if ($address['country'] == 'AU') {
-    $format['locality_block']['administrative_area']['#options'] = array(
-      '' => t('--'),
-      'ACT' => t('Australian Capital Territory'),
-      'NSW' => t('New South Wales'),
-      'NT' => t('Northern Territory'),
-      'QLD' => t('Queensland'),
-      'SA' => t('South Australia'),
-      'TAS' => t('Tasmania'),
-      'VIC' => t('Victoria'),
-      'WA' => t('Western Australia'),
-    );
   }
   else if ($address['country'] == 'NZ') {
     $format['locality_block']['locality']['#title'] = ('Town/City');
@@ -431,26 +200,6 @@ function addressfield_format_address_generate(&$format, $address, $context = arr
     $format['locality_block']['administrative_area']['#render_option_value'] = TRUE;
     $format['locality_block']['administrative_area']['#title'] = t('Region');
     $format['locality_block']['administrative_area']['#required'] = FALSE;
-    $format['locality_block']['administrative_area']['#options'] = array(
-      ''   => t('--'),
-      'AUK' => t('Auckland'),
-      'BOP' => t('Bay of Plenty'),
-      'CAN' => t('Canterbury'),
-      'HKB' => t("Hawke's Bay"),
-      'MWT' => t('Manawatu-Wanganui'),
-      'NTL' => t('Northland'),
-      'OTA' => t('Otago'),
-      'STL' => t('Southland'),
-      'TKI' => t('Taranaki'),
-      'WKO' => t('Waikato'),
-      'WGN' => t('Wellington'),
-      'WTC' => t('West Coast'),
-      'GIS' => t('Gisborne District'),
-      'MBH' => t('Marlborough District'),
-      'NSN' => t('Nelson'),
-      'TAS' => t('Tasman District'),
-      'CIT' => t('Chatham Islands Territory'),
-    );
   }
 
   // Those countries tend to put the postal code after the locality.
@@ -542,3 +291,304 @@ function addressfield_format_address_generate(&$format, $address, $context = arr
     }
   }
 }
+
+
+/**
+ * Generates and array of available administrative areas for countries.
+ * @param string $country_code
+ * @return array
+ */
+function country_administrative_areas($country_code) {
+  switch ($country_code) {
+    case 'US':
+      $areas = array(
+        ''   => t('--'),
+        'AL' => t('Alabama'),
+        'AK' => t('Alaska'),
+        'AZ' => t('Arizona'),
+        'AR' => t('Arkansas'),
+        'CA' => t('California'),
+        'CO' => t('Colorado'),
+        'CT' => t('Connecticut'),
+        'DE' => t('Delaware'),
+        'DC' => t('District Of Columbia'),
+        'FL' => t('Florida'),
+        'GA' => t('Georgia'),
+        'HI' => t('Hawaii'),
+        'ID' => t('Idaho'),
+        'IL' => t('Illinois'),
+        'IN' => t('Indiana'),
+        'IA' => t('Iowa'),
+        'KS' => t('Kansas'),
+        'KY' => t('Kentucky'),
+        'LA' => t('Louisiana'),
+        'ME' => t('Maine'),
+        'MD' => t('Maryland'),
+        'MA' => t('Massachusetts'),
+        'MI' => t('Michigan'),
+        'MN' => t('Minnesota'),
+        'MS' => t('Mississippi'),
+        'MO' => t('Missouri'),
+        'MT' => t('Montana'),
+        'NE' => t('Nebraska'),
+        'NV' => t('Nevada'),
+        'NH' => t('New Hampshire'),
+        'NJ' => t('New Jersey'),
+        'NM' => t('New Mexico'),
+        'NY' => t('New York'),
+        'NC' => t('North Carolina'),
+        'ND' => t('North Dakota'),
+        'OH' => t('Ohio'),
+        'OK' => t('Oklahoma'),
+        'OR' => t('Oregon'),
+        'PA' => t('Pennsylvania'),
+        'RI' => t('Rhode Island'),
+        'SC' => t('South Carolina'),
+        'SD' => t('South Dakota'),
+        'TN' => t('Tennessee'),
+        'TX' => t('Texas'),
+        'UT' => t('Utah'),
+        'VT' => t('Vermont'),
+        'VA' => t('Virginia'),
+        'WA' => t('Washington'),
+        'WV' => t('West Virginia'),
+        'WI' => t('Wisconsin'),
+        'WY' => t('Wyoming'),
+        ' ' => t('--'),
+        'AA' => t('Armed Forces (Americas)'),
+        'AE' => t('Armed Forces (Europe, Canada, Middle East, Africa)'),
+        'AP' => t('Armed Forces (Pacific)'),
+        'AS' => t('American Samoa'),
+        'FM' => t('Federated States of Micronesia'),
+        'GU' => t('Guam'),
+        'MH' => t('Marshall Islands'),
+        'MP' => t('Northern Mariana Islands'),
+        'PW' => t('Palau'),
+        'PR' => t('Puerto Rico'),
+        'VI' => t('Virgin Islands'),
+      );
+      break;
+    case 'IT':
+      $areas = array(
+        '' => t('--'),
+        'AG' => t('Agrigento'),
+        'AL' => t('Alessandria'),
+        'AN' => t('Ancona'),
+        'AO' => t("Valle d'Aosta/Vallée d'Aoste"),
+        'AP' => t('Ascoli Piceno'),
+        'AQ' => t("L'Aquila"),
+        'AR' => t('Arezzo'),
+        'AT' => t('Asti'),
+        'AV' => t('Avellino'),
+        'BA' => t('Bari'),
+        'BG' => t('Bergamo'),
+        'BI' => t('Biella'),
+        'BL' => t('Belluno'),
+        'BN' => t('Benevento'),
+        'BO' => t('Bologna'),
+        'BR' => t('Brindisi'),
+        'BS' => t('Brescia'),
+        'BT' => t('Barletta-Andria-Trani'),
+        'BZ' => t('Bolzano/Bozen'),
+        'CA' => t('Cagliari'),
+        'CB' => t('Campobasso'),
+        'CE' => t('Caserta'),
+        'CH' => t('Chieti'),
+        'CI' => t('Carbonia-Iglesias'),
+        'CL' => t('Caltanissetta'),
+        'CN' => t('Cuneo'),
+        'CO' => t('Como'),
+        'CR' => t('Cremona'),
+        'CS' => t('Cosenza'),
+        'CT' => t('Catania'),
+        'CZ' => t('Catanzaro'),
+        'EN' => t('Enna'),
+        'FC' => t('Forlì-Cesena'),
+        'FE' => t('Ferrara'),
+        'FG' => t('Foggia'),
+        'FI' => t('Firenze'),
+        'FM' => t('Fermo'),
+        'FR' => t('Frosinone'),
+        'GE' => t('Genova'),
+        'GO' => t('Gorizia'),
+        'GR' => t('Grosseto'),
+        'IM' => t('Imperia'),
+        'IS' => t('Isernia'),
+        'KR' => t('Crotone'),
+        'LC' => t('Lecco'),
+        'LE' => t('Lecce'),
+        'LI' => t('Livorno'),
+        'LO' => t('Lodi'),
+        'LT' => t('Latina'),
+        'LU' => t('Lucca'),
+        'MB' => t('Monza e Brianza'),
+        'MC' => t('Macerata'),
+        'ME' => t('Messina'),
+        'MI' => t('Milano'),
+        'MN' => t('Mantova'),
+        'MO' => t('Modena'),
+        'MS' => t('Massa-Carrara'),
+        'MT' => t('Matera'),
+        'NA' => t('Napoli'),
+        'NO' => t('Novara'),
+        'NU' => t('Nuoro'),
+        'OG' => t('Ogliastra'),
+        'OR' => t('Oristano'),
+        'OT' => t('Olbia-Tempio'),
+        'PA' => t('Palermo'),
+        'PC' => t('Piacenza'),
+        'PD' => t('Padova'),
+        'PE' => t('Pescara'),
+        'PG' => t('Perugia'),
+        'PI' => t('Pisa'),
+        'PN' => t('Pordenone'),
+        'PO' => t('Prato'),
+        'PR' => t('Parma'),
+        'PT' => t('Pistoia'),
+        'PU' => t('Pesaro e Urbino'),
+        'PV' => t('Pavia'),
+        'PZ' => t('Potenza'),
+        'RA' => t('Ravenna'),
+        'RC' => t('Reggio di Calabria'),
+        'RE' => t("Reggio nell'Emilia"),
+        'RG' => t('Ragusa'),
+        'RI' => t('Rieti'),
+        'RM' => t('Roma'),
+        'RN' => t('Rimini'),
+        'RO' => t('Rovigo'),
+        'SA' => t('Salerno'),
+        'SI' => t('Siena'),
+        'SO' => t('Sondrio'),
+        'SP' => t('La Spezia'),
+        'SR' => t('Siracusa'),
+        'SS' => t('Sassari'),
+        'SV' => t('Savona'),
+        'TA' => t('Taranto'),
+        'TE' => t('Teramo'),
+        'TN' => t('Trento'),
+        'TO' => t('Torino'),
+        'TP' => t('Trapani'),
+        'TR' => t('Terni'),
+        'TS' => t('Trieste'),
+        'TV' => t('Treviso'),
+        'UD' => t('Udine'),
+        'VA' => t('Varese'),
+        'VB' => t('Verbano-Cusio-Ossola'),
+        'VC' => t('Vercelli'),
+        'VE' => t('Venezia'),
+        'VI' => t('Vicenza'),
+        'VR' => t('Verona'),
+        'VS' => t('Medio Campidano'),
+        'VT' => t('Viterbo'),
+        'VV' => t('Vibo Valentia'),
+      );
+      break;
+    case 'BR':
+      $areas = array(
+        '' => t('--'),
+        'AC' => t('Acre'),
+        'AL' => t('Alagoas'),
+        'AM' => t('Amazonas'),
+        'AP' => t('Amapá'),
+        'BA' => t('Bahia'),
+        'CE' => t('Ceará'),
+        'DF' => t('Distrito Federal'),
+        'ES' => t('Espírito Santo'),
+        'GO' => t('Goiás'),
+        'MA' => t('Maranhão'),
+        'MG' => t('Minas Gerais'),
+        'MS' => t('Mato Grosso do Sul'),
+        'MT' => t('Mato Grosso'),
+        'PA' => t('Pará'),
+        'PB' => t('Paraíba'),
+        'PE' => t('Pernambuco'),
+        'PI' => t('Piauí'),
+        'PR' => t('Paraná'),
+        'RJ' => t('Rio de Janeiro'),
+        'RN' => t('Rio Grande do Norte'),
+        'RO' => t('Rondônia'),
+        'RR' => t('Roraima'),
+        'RS' => t('Rio Grande do Sul'),
+        'SC' => t('Santa Catarina'),
+        'SE' => t('Sergipe'),
+        'SP' => t('São Paulo'),
+        'TO' => t('Tocantins'),
+      );
+      break;
+    case 'CA':
+      $areas = array(
+        '' => t('--'),
+        'AB' => t('Alberta'),
+        'BC' => t('British Columbia'),
+        'MB' => t('Manitoba'),
+        'NB' => t('New Brunswick'),
+        'NL' => t('Newfoundland and Labrador'),
+        'NT' => t('Northwest Territories'),
+        'NS' => t('Nova Scotia'),
+        'NU' => t('Nunavut'),
+        'ON' => t('Ontario'),
+        'PE' => t('Prince Edward Island'),
+        'QC' => t('Quebec'),
+        'SK' => t('Saskatchewan'),
+        'YT' => t('Yukon Territory'),
+      );
+      break;
+    case 'AU':
+      $areas = array(
+        '' => t('--'),
+        'ACT' => t('Australian Capital Territory'),
+        'NSW' => t('New South Wales'),
+        'NT' => t('Northern Territory'),
+        'QLD' => t('Queensland'),
+        'SA' => t('South Australia'),
+        'TAS' => t('Tasmania'),
+        'VIC' => t('Victoria'),
+        'WA' => t('Western Australia'),
+      );
+      break;
+    case 'NZ':
+      $areas = array(
+        ''   => t('--'),
+        'AUK' => t('Auckland'),
+        'BOP' => t('Bay of Plenty'),
+        'CAN' => t('Canterbury'),
+        'HKB' => t("Hawke's Bay"),
+        'MWT' => t('Manawatu-Wanganui'),
+        'NTL' => t('Northland'),
+        'OTA' => t('Otago'),
+        'STL' => t('Southland'),
+        'TKI' => t('Taranaki'),
+        'WKO' => t('Waikato'),
+        'WGN' => t('Wellington'),
+        'WTC' => t('West Coast'),
+        'GIS' => t('Gisborne District'),
+        'MBH' => t('Marlborough District'),
+        'NSN' => t('Nelson'),
+        'TAS' => t('Tasman District'),
+        'CIT' => t('Chatham Islands Territory'),
+      );
+      break;
+    default:
+      $areas = array();
+  }
+
+  return $areas;
+}
+
+
+/**
+ * Gets administrative area name by the area code.
+ * @param string $country_code
+ * @param string $administrative_area_code
+ * @return string|bool
+ *   Returns administrative area name if the match exists, and FALSE otherwise
+ */
+function administrative_area_name_by_code($country_code, $administrative_area_code) {
+  $available_areas = country_administrative_areas($country_code);
+
+  if (!empty($available_areas)) {
+    return $available_areas[$administrative_area_code];
+  }
+
+  return FALSE;
+}
\ No newline at end of file
