diff --git a/includes/twitter_bootstrap.inc b/includes/twitter_bootstrap.inc
index be302ef..a291091 100644
--- a/includes/twitter_bootstrap.inc
+++ b/includes/twitter_bootstrap.inc
@@ -79,13 +79,16 @@ function twitter_bootstrap_menu_navigation_links($tree, $lvl = 0) {
 
   if(count($tree) > 0) {
     foreach($tree as $id => $item) {
-      $new_item = array('title' => $item['link']['title'], 'link_path' => $item['link']['link_path'], 'href' => $item['link']['href']);
+      // Only work on enabled links
+      if (empty($item['link']['hidden'])) {
+        $new_item = array('title' => $item['link']['title'], 'link_path' => $item['link']['link_path'], 'href' => $item['link']['href']);
     
-      // Dont do drugs and don't do any levels deeper then 1
-      if($lvl < 1)
-      $new_item['below'] = twitter_bootstrap_menu_navigation_links($item['below'], $lvl+1);
+        // Dont do drugs and don't do any levels deeper then 1
+        if($lvl < 1)
+        $new_item['below'] = twitter_bootstrap_menu_navigation_links($item['below'], $lvl+1);
     
-      $result['menu-'. $item['link']['mlid']] = $new_item;
+        $result['menu-'. $item['link']['mlid']] = $new_item;
+      }
     }
   }
   
